docs/proposals/templates.md

```diff
@@ -0,0 +1,550 @@
+      "required": true,
+      "type" :"string"
+    },
+    {
+      "name": "MONGODB_DATABASE",
+      "description": "Database name",
+      "value": "sampledb",
+      "required": true,
+      "type" :"string"      
+    },
+    {
+      "name": "REPLICA_COUNT",
+      "description": "Number of mongo replicas to run",
+      "value": "1",
+      "required": true,
+      "type" :"int"      
+    }
+  ]
+}
+```
+
+### API Endpoints
+
+* **/processedTemplates** - when a template is POSTed to this endpoint, all parameters in the template are processed, values 
+generated as necessary, and substituted into appropriate locations in the object definitions.  Validation is performed to ensure
+required parameters have a value (either supplied in the original template or generated by a defined generator).  In addition 
+labels defined in the template are applied to the object definitions.  Finally the customized template (still a `Template` object) 
+is returned to the caller.
+
+Performing parameter substitution on the server side has the benefit of centralizing the processing so that new clients of 
+k8s, such as IDEs, CI systems, Web consoles, etc, do not need to reimplement template processing or embed the k8s binary.
+Instead they can invoke the k8s api directly.
+
+This resource can also support a subresource "/processedTemplates/templatename".  This resource would accept just a Parameters object
+and would process the template stored in the cluster as "templatename".  The processed result would be returned.
+
+* **/templates** - the REST storage resource for storing and retrieving template objects, scoped within a namespace.
+
+Storing templates within k8s has the benefit of enabling template sharing and securing via the same roles/resources
+that are used to provide access control to other cluster resoures.  It also enables sophisticated service catalog
+flows in which selecting a service from a catalog results in a new instantiation of that service.  (This is not the
+only way to implement such a flow, but it does provide a useful level of integration).
+
+### Workflow
+
+#### Template Instantiation
+
+Given a well-formed template, a client will
+
+1. Optionally set an explicit `value` for any parameter values the user wishes to explicitly set (eg non-generated parameters for 
+   which the default value is not desired)
+2. Submit the new template object to the `/processedTemplates` api endpoint
+
+The api endpoint will then:
+
+1. Validate the template including confirming “required” parameters have an explicit value.
+2. Walk each api object in the template.
+3. Adding all labels defined in the template’s ObjectLabels field.
+4. For each field, check if the value matches a parameter name and if so, set the value of the field to the value of the parameter.
+  * Partial substitutions are accepted, such as `SOME_$(PARAM)` which would be transformed into `SOME_XXXX` where `XXXX` is the value
```

Comment:
I would think a reasonable assumption is that template substitution occurs first and therefore takes precedence over env expansion, rather than being in conflict.

Checking whether a given $(VAL) could also represent an env variable to be expanded is going to complicate the processing quite a bit, i'd think (though i'm not very familiar w/ the env expansion logic)